### ISSUE TITLE
Flake caused by unfortunately timed drop of watches?

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#20599]
I'm tracking down an e2e flake here: https://github.com/kubernetes/kubernetes/pull/15938#issuecomment-179480200

I think what happened was the the kube-controller-manager "missed" a load-balancer change, because just after it was created we got this:

```
I0203 21:14:48.708240       7 reflector.go:366] /usr/local/go/src/runtime/asm_amd64.s:1721: Watch close - *api.Service total 1 items received
I0203 21:14:48.708293       7 reflector.go:294] /usr/local/go/src/runtime/asm_amd64.s:1721: next resync planned for time.Time{sec:63590130890, nsec:705171438, loc:(*time.Location)(0x24e8c40)}, forcing now
I0203 21:14:48.806812       7 reflector.go:366] /usr/local/go/src/runtime/asm_amd64.s:1721: Watch close - *extensions.Job total 1 items received
I0203 21:14:48.806867       7 reflector.go:294] /usr/local/go/src/runtime/asm_amd64.s:1721: next resync planned for time.Time{sec:63590130890, nsec:804435543, loc:(*time.Location)(0x24e8c40)}, forcing now
I0203 21:14:48.892173       7 reflector.go:366] /usr/local/go/src/runtime/asm_amd64.s:1721: Watch close - *extensions.DaemonSet total 0 items received
I0203 21:14:48.892231       7 reflector.go:294] /usr/local/go/src/runtime/asm_amd64.s:1721: next resync planned for time.Time{sec:63590130890, nsec:889461997, loc:(*time.Location)(0x24e8c40)}, forcing now
I0203 21:14:48.910751       7 reflector.go:366] /usr/local/go/src/runtime/asm_amd64.s:1721: Watch close - *extensions.Deployment total 0 items received
I0203 21:14:48.910821       7 reflector.go:294] /usr/local/go/src/runtime/asm_amd64.s:1721: next resync planned for time.Time{sec:63590130890, nsec:908694152, loc:(*time.Location)(0x24e8c40)}, forcing now
I0203 21:14:49.437044       7 reflector.go:366] /usr/local/go/src/runtime/asm_amd64.s:1721: Watch close - *api.ReplicationController total 1 items received
I0203 21:14:49.437083       7 reflector.go:294] /usr/local/go/src/runtime/asm_amd64.s:1721: next resync planned for time.Time{sec:63590130891, nsec:431461876, loc:(*time.Location)(0x24e8c40)}, forcing now
```

The e2e test had at approximately the same time issued a change on mutability-service-test, changing ServiceType from LoadBalancer to ClusterIP, but the LB was not deleted.  We did see this:

```
I0203 21:14:50.443567       7 controller_utils.go:152] Controller e2e-tests-services-gb3yh/webserver either never recorded expectations, or the ttl expired.
I0203 21:14:50.443585       7 replication_controller.go:410] Finished syncing controller "e2e-tests-services-gb3yh/webserver" (20.344Âµs)
I0203 21:14:51.919362       7 servicecontroller.go:338] Not persisting unchanged LoadBalancerStatus to registry.
I0203 21:14:51.919439       7 servicecontroller.go:278] LB already exists and doesn't need update for service e2e-tests-services-gb3yh/mutability-service-test
I0203 21:14:51.919460       7 servicecontroller.go:278] LB already exists and doesn't need update for service e2e-tests-services-gb3yh/mutability-service-test
```

I suspect the logic here is broken after a resync.  But I don't really understand why the watches would all close, or where to look for a bug here.  My hypothesis is that either the watch is not resuming correctly, or the LB controller code is not doing a comparison correctly.

EDIT: I now think that we're actually forcing a resync, not that the connections are closing (I think they're correlated because there's no randomness in the intervals and we're likely to start them all at the same time I believe.)  And I think it looks like a known etcd issue with a watch not working properly after a list (https://github.com/coreos/etcd/issues/2854#issuecomment-105329971).

**Comment by lavalamp** [comment]
Correlated watch closes are problematic, so we should figure out why they all closed at the same time.

It sounds like there's a bug in the service controller, since the whole point of the list/watch cycle is that you never need to miss an update.

**Comment by lavalamp** [comment]
Kris, can you take a look at this? I'm very interested to know why the watches all closed at the same time.

**Comment by lavalamp** [comment]
@justinsb can you post a link to the full logs?

**Comment by justinsb** [comment]
The comment above links to them, but copying it here...

GCE e2e build/test **failed** for commit f61a5d0400218ea8334b75edd42831a1cbeed262.
* [Build Log](https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/pull/15938/kubernetes-pull-build-test-e2e-gce/26565/build-log.txt)
* [Test Artifacts](https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/15938/kubernetes-pull-build-test-e2e-gce/26565/artifacts/)
* [Internal Jenkins Results](http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//26565)


Have been staring at this one and still haven't figured out what is going on.

**Comment by justinsb** [comment]
Here's what I think are the most relevant lines from the logs...

First watch failure
```
I0203 21:14:37.466072       7 reflector.go:366] pkg/controller/service/servicecontroller.go:138: Watch close - *api.Service total 20 items received
I0203 21:14:37.466120       7 reflector.go:294] pkg/controller/service/servicecontroller.go:138: next resync planned for time.Time{sec:63590130879, nsec:462103355, loc:(*time.Location)(0x24e8c40)}, forcing now
```

Sync on service
```
I0203 21:14:38.468362       7 servicecontroller.go:223] Got new Sync delta for service: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:mutability-service-test GenerateName: Namespace:e2e-tests-services-gb3yh SelfLink:/api/v1/namespaces/e2e-tests-services-gb3yh/services/mutability-service-test UID:ab2f7739-caba-11e5-ba15-42010af00002 ResourceVersion:8815 Generation:0 CreationTimestamp:2016-02-03 21:11:18 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[]} Spec:{Type:LoadBalancer Ports:[{Name: Protocol:TCP Port:19482 TargetPort:{Type:0 IntVal:80 StrVal:} NodePort:31511}] Selector:map[testid:mutability-service-test-ab2eda1f-caba-11e5-8da3-42010af00007] ClusterIP:10.0.195.70 ExternalIPs:[] LoadBalancerIP: SessionAffinity:None} Status:{LoadBalancer:{Ingress:[{IP:104.197.185.91 Hostname:}]}}}
```

Update on service (but note that Type still = LoadBalancer, and that we haven't seen any processing of the sync event yet)..
```
I0203 21:14:38.470198       7 servicecontroller.go:223] Got new Updated delta for service: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:mutability-service-test GenerateName: Namespace:e2e-tests-services-gb3yh SelfLink:/api/v1/namespaces/e2e-tests-services-gb3yh/services/mutability-service-test UID:ab2f7739-caba-11e5-ba15-42010af00002 ResourceVersion:8815 Generation:0 CreationTimestamp:2016-02-03 21:11:18 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[]} Spec:{Type:LoadBalancer Ports:[{Name: Protocol:TCP Port:19482 TargetPort:{Type:0 IntVal:80 StrVal:} NodePort:31511}] Selector:map[testid:mutability-service-test-ab2eda1f-caba-11e5-8da3-42010af00007] ClusterIP:10.0.195.70 ExternalIPs:[] LoadBalancerIP: SessionAffinity:None} Status:{LoadBalancer:{Ingress:[{IP:104.197.185.91 Hostname:}]}}}
```

Watch fail 2:
```
I0203 21:14:48.708240       7 reflector.go:366] /usr/local/go/src/runtime/asm_amd64.s:1721: Watch close - *api.Service total 1 items received
I0203 21:14:48.708293       7 reflector.go:294] /usr/local/go/src/runtime/asm_amd64.s:1721: next resync planned for time.Time{sec:63590130890, nsec:705171438, loc:(*time.Location)(0x24e8c40)}, forcing now
```

Processing of events - both no-ops:
```
I0203 21:14:51.919439       7 servicecontroller.go:278] LB already exists and doesn't need update for service e2e-tests-services-gb3yh/mutability-service-test
I0203 21:14:51.919460       7 servicecontroller.go:278] LB already exists and doesn't need update for service e2e-tests-services-gb3yh/mutability-service-test
```

**Comment by justinsb** [comment]
A different potential problem (I think): in `pkg/controller/service/servicecontroller.go` we spawn 10 worker goroutines, which pop from the serviceQueue.  We won't process the same item concurrently, because of the `cachedService.mu.Lock` in processDelta.  But I don't see any guarantee that the order in which the goroutines win the race to acquire the `cachedService.mu` lock is the same as the order in which we pop them from the serviceQueue.  So out-of-order processing is possible.

Based on the logs, I don't believe that is happening here (although we do have the conditions for it to happen, with a prior change holding the lock so 2 changes are queued up).  I think the fact that both the Sync and Update messages were the same (same ResourceVersion) is the most suspicious thing here.

**Comment by justinsb** [comment]
Another very suspicious thing; GetByKey returns an serviceCache object, instead of an *api.Service:

https://github.com/kubernetes/kubernetes/blob/843c11e06ac7dcdadb6ea47df53367a58a415085/pkg/controller/service/servicecontroller.go#L409

Again though, I don't think it's a problem here - it only seems to be called by `DeltaFIFO::Replace` in the case of a deletion not already in the queue, and that isn't what we're seeing here.

**Comment by justinsb** [comment]
Sorry to bombard, but here's what I think is the "smoking gun"... grepping in the kube-controller-manager log for `api/v1/namespaces/e2e-tests-services-gb3yh/services/mutability-service-test`, which brings up just the changes...

```
I0203 21:11:18.945101       7 servicecontroller.go:223] Got new Added delta for service: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:mutability-service-test GenerateName: Namespace:e2e-tests-services-gb3yh SelfLink:/api/v1/namespaces/e2e-tests-services-gb3yh/services/mutability-service-test UID:ab2f7739-caba-11e5-ba15-42010af00002 ResourceVersion:8550 Generation:0 CreationTimestamp:2016-02-03 21:11:18 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[]} Spec:{Type:ClusterIP Ports:[{Name: Protocol:TCP Port:80 TargetPort:{Type:0 IntVal:80 StrVal:} NodePort:0}] Selector:map[testid:mutability-service-test-ab2eda1f-caba-11e5-8da3-42010af00007] ClusterIP:10.0.195.70 ExternalIPs:[] LoadBalancerIP: SessionAffinity:None} Status:{LoadBalancer:{Ingress:[]}}}
I0203 21:11:23.979524       7 servicecontroller.go:223] Got new Updated delta for service: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:mutability-service-test GenerateName: Namespace:e2e-tests-services-gb3yh SelfLink:/api/v1/namespaces/e2e-tests-services-gb3yh/services/mutability-service-test UID:ab2f7739-caba-11e5-ba15-42010af00002 ResourceVersion:8585 Generation:0 CreationTimestamp:2016-02-03 21:11:18 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[]} Spec:{Type:NodePort Ports:[{Name: Protocol:TCP Port:80 TargetPort:{Type:0 IntVal:80 StrVal:} NodePort:31510}] Selector:map[testid:mutability-service-test-ab2eda1f-caba-11e5-8da3-42010af00007] ClusterIP:10.0.195.70 ExternalIPs:[] LoadBalancerIP: SessionAffinity:None} Status:{LoadBalancer:{Ingress:[]}}}
I0203 21:11:26.003069       7 servicecontroller.go:223] Got new Updated delta for service: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:mutability-service-test GenerateName: Namespace:e2e-tests-services-gb3yh SelfLink:/api/v1/namespaces/e2e-tests-services-gb3yh/services/mutability-service-test UID:ab2f7739-caba-11e5-ba15-42010af00002 ResourceVersion:8599 Generation:0 CreationTimestamp:2016-02-03 21:11:18 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[]} Spec:{Type:LoadBalancer Ports:[{Name: Protocol:TCP Port:80 TargetPort:{Type:0 IntVal:80 StrVal:} NodePort:31510}] Selector:map[testid:mutability-service-test-ab2eda1f-caba-11e5-8da3-42010af00007] ClusterIP:10.0.195.70 ExternalIPs:[] LoadBalancerIP: SessionAffinity:None} Status:{LoadBalancer:{Ingress:[]}}}
I0203 21:12:23.419457       7 servicecontroller.go:223] Got new Updated delta for service: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:mutability-service-test GenerateName: Namespace:e2e-tests-services-gb3yh SelfLink:/api/v1/namespaces/e2e-tests-services-gb3yh/services/mutability-service-test UID:ab2f7739-caba-11e5-ba15-42010af00002 ResourceVersion:8806 Generation:0 CreationTimestamp:2016-02-03 21:11:18 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[]} Spec:{Type:LoadBalancer Ports:[{Name: Protocol:TCP Port:80 TargetPort:{Type:0 IntVal:80 StrVal:} NodePort:31510}] Selector:map[testid:mutability-service-test-ab2eda1f-caba-11e5-8da3-42010af00007] ClusterIP:10.0.195.70 ExternalIPs:[] LoadBalancerIP: SessionAffinity:None} Status:{LoadBalancer:{Ingress:[{IP:104.197.185.91 Hostname:}]}}}
I0203 21:12:26.126412       7 servicecontroller.go:223] Got new Updated delta for service: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:mutability-service-test GenerateName: Namespace:e2e-tests-services-gb3yh SelfLink:/api/v1/namespaces/e2e-tests-services-gb3yh/services/mutability-service-test UID:ab2f7739-caba-11e5-ba15-42010af00002 ResourceVersion:8811 Generation:0 CreationTimestamp:2016-02-03 21:11:18 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[]} Spec:{Type:LoadBalancer Ports:[{Name: Protocol:TCP Port:80 TargetPort:{Type:0 IntVal:80 StrVal:} NodePort:31511}] Selector:map[testid:mutability-service-test-ab2eda1f-caba-11e5-8da3-42010af00007] ClusterIP:10.0.195.70 ExternalIPs:[] LoadBalancerIP: SessionAffinity:None} Status:{LoadBalancer:{Ingress:[{IP:104.197.185.91 Hostname:}]}}}
I0203 21:12:28.147332       7 servicecontroller.go:223] Got new Updated delta for service: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:mutability-service-test GenerateName: Namespace:e2e-tests-services-gb3yh SelfLink:/api/v1/namespaces/e2e-tests-services-gb3yh/services/mutability-service-test UID:ab2f7739-caba-11e5-ba15-42010af00002 ResourceVersion:8815 Generation:0 CreationTimestamp:2016-02-03 21:11:18 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[]} Spec:{Type:LoadBalancer Ports:[{Name: Protocol:TCP Port:19482 TargetPort:{Type:0 IntVal:80 StrVal:} NodePort:31511}] Selector:map[testid:mutability-service-test-ab2eda1f-caba-11e5-8da3-42010af00007] ClusterIP:10.0.195.70 ExternalIPs:[] LoadBalancerIP: SessionAffinity:None} Status:{LoadBalancer:{Ingress:[{IP:104.197.185.91 Hostname:}]}}}
I0203 21:14:38.468362       7 servicecontroller.go:223] Got new Sync delta for service: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:mutability-service-test GenerateName: Namespace:e2e-tests-services-gb3yh SelfLink:/api/v1/namespaces/e2e-tests-services-gb3yh/services/mutability-service-test UID:ab2f7739-caba-11e5-ba15-42010af00002 ResourceVersion:8815 Generation:0 CreationTimestamp:2016-02-03 21:11:18 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[]} Spec:{Type:LoadBalancer Ports:[{Name: Protocol:TCP Port:19482 TargetPort:{Type:0 IntVal:80 StrVal:} NodePort:31511}] Selector:map[testid:mutability-service-test-ab2eda1f-caba-11e5-8da3-42010af00007] ClusterIP:10.0.195.70 ExternalIPs:[] LoadBalancerIP: SessionAffinity:None} Status:{LoadBalancer:{Ingress:[{IP:104.197.185.91 Hostname:}]}}}
I0203 21:14:38.470198       7 servicecontroller.go:223] Got new Updated delta for service: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:mutability-service-test GenerateName: Namespace:e2e-tests-services-gb3yh SelfLink:/api/v1/namespaces/e2e-tests-services-gb3yh/services/mutability-service-test UID:ab2f7739-caba-11e5-ba15-42010af00002 ResourceVersion:8815 Generation:0 CreationTimestamp:2016-02-03 21:11:18 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[]} Spec:{Type:LoadBalancer Ports:[{Name: Protocol:TCP Port:19482 TargetPort:{Type:0 IntVal:80 StrVal:} NodePort:31511}] Selector:map[testid:mutability-service-test-ab2eda1f-caba-11e5-8da3-42010af00007] ClusterIP:10.0.195.70 ExternalIPs:[] LoadBalancerIP: SessionAffinity:None} Status:{LoadBalancer:{Ingress:[{IP:104.197.185.91 Hostname:}]}}}
```

We expect to observe as the final change a change from Type:LoadBalancer -> Type:ClusterIP (and NodePort=0).  But we don't.  We do however see repeated Updated messages with the same ResourceVersion.

Looking just at the message types are the resource versions...


```
Added ResourceVersion:8550 
Updated ResourceVersion:8585
Updated ResourceVersion:8599
Updated ResourceVersion:8806
Updated ResourceVersion:8811
Updated ResourceVersion:8815
Sync ResourceVersion:8815
Updated ResourceVersion:8815
```

So everything is going great, until we hit that Sync, and then we get a spurious update with the same version, and not the update we expect when we change the service type.

**Comment by justinsb** [comment]
OK, one last thought: https://github.com/coreos/etcd/issues/2854#issuecomment-105329971

I think this is _exactly_ what we're seeing.  We get a spurious update, and then we get nothing else.  The e2e test is waiting for the service to change before making additional changes.

@bprashanth do you think this could be what we're seeing here?

**Comment by bprashanth** [comment]
I haven't completely parsed your debugging, but I noticed in parallel that we might be dropping updates https://github.com/kubernetes/kubernetes/issues/20547. Have you ruled out service controller itself dropping things? I'm adding events in https://github.com/kubernetes/kubernetes/issues/20607 to help

**Comment by justinsb** [comment]
@bprashanth I definitely haven't ruled out service controller itself dropping things.  I think I've found a few things in the code that are suspicious.  But the logs that I'm looking at don't support those things being a problem.

Instead, they showed a Sync event coming at almost exactly the same time as the e2e test makes a change, which then gets dropped.  After the Sync, I saw an Update event with the same ResourceVersion, and then nothing.

This seems to match that bug on etcd, where a new watch will hang unless further changes are made.  Is it your understanding that this bug is still open in etcd?
